infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java

```diff
@@ -0,0 +1,219 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class allows mount user SSH Keys as <a
+ * href="https://kubernetes.io/docs/concepts/configuration/secret">Kubernetes Secrets</a> on each
+ * workspace containers. <br>
+ * <strong>How it works</strong> <br>
+ * When starting a workspace, we will grep all SSH Keys registered for VCS and create K8s Secret the
+ * to them. Then secrets will be mounted on each container by path
+ * '/etc/ssh/{sshKeyName}/ssh-privatekey' as a file. <a
+ * href=https://github.com/kubernetes/kubernetes/blob/7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/pkg/apis/core/types.go#L4458">Secret
+ * Type for SSH keys</a> Also will be created and mounted ConfigMap with SSH settings to the
+ * '/etc/ssh/ssh_config'.
+ *
+ * @author Vitalii Parfonov
+ */
+public class VcsSshKeysProvisioner implements ConfigurationProvisioner<KubernetesEnvironment> {
+
+  // SecretTypeSSHAuth contains data needed for SSH authentication.
+  // Required field:
+  // - Secret.Data["ssh-privatekey"] - private SSH key needed for authentication
+  private static final String SECRET_TYPE_SSH = "kubernetes.io/ssh-auth";
+
+  // SSHAuthPrivateKey is the key of the required SSH private key for SecretTypeSSHAuth secrets
+  private static final String SSH_PRIVATE_KEY = "ssh-privatekey";
+
+  private static final String SSH_BASE_CONFIG_PATH = "/etc/ssh/";
+
+  private static final String SSH_CONFIG = "ssh_config";
+
+  private static final String SSH_CONFIG_PATH = SSH_BASE_CONFIG_PATH + SSH_CONFIG;
+
+  private static final Logger LOG = LoggerFactory.getLogger(VcsSshKeysProvisioner.class);
+
+  private final SshManager sshManager;
+
+  @Inject
+  public VcsSshKeysProvisioner(SshManager sshManager) {
+    this.sshManager = sshManager;
+  }
+
+  @Override
+  @Traced
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);
+
+    StringBuilder sshConfig = new StringBuilder();
+    Map<String, String> map = new HashMap<>();
+
+    try {
+      List<SshPairImpl> sshPairs = sshManager.getPairs(identity.getOwnerId(), "vcs");
+      if (!sshPairs.isEmpty()) {
+        for (SshPair sshPair : sshPairs) {
+          if (isNullOrEmpty(sshPair.getName()) || isNullOrEmpty(sshPair.getPrivateKey())) {
+            continue;
+          }
+          Secret secret =
+              new SecretBuilder()
+                  .addToData(
+                      SSH_PRIVATE_KEY,
+                      Base64.getEncoder().encodeToString(sshPair.getPrivateKey().getBytes()))
+                  .withType(SECRET_TYPE_SSH)
+                  .withNewMetadata()
+                  .withName(getValidNameForSecret(sshPair.getName()))
+                  .endMetadata()
+                  .build();
+
+          k8sEnv.getSecrets().put(secret.getMetadata().getName(), secret);
+
+          if (!sshPairs.isEmpty()) {
```

Comment:
This check seems redundant for me, isn't we already checked this in line 96?

Reply:
yes. thanks